Question 1:

Load the dataset and return the dataframe.

In [205]:
import pandas as pd
import numpy as np
df = pd.read_csv("heart.csv")
df

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  ca  thal  target
0     63    1   3       145   233    1        0      150      0      2.3      0   0     1       1
1     37    1   2       130   250    0        1      187      0      3.5      0   0     2       1
2     41    0   1       130   204    0        0      172      0      1.4      2   0     2       1
3     56    1   1       120   236    0        1      178      0      0.8      2   0     2       1
4     57    0   0       120   354    0        1      163      1      0.6      2   0     2       1
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...    ...  ..   ...     ...
298   57    0   0       140   241    0        1      123      1      0.2      1   0     3       0
299   45    1   3       110   264    0        1      132      0      1.2      1   0     3       0
300   68    1   0       144   193    1        1      141      0      3.4      1   2     3       0
301   57    1   0       130   131    0        1      115      1      1.2      1   1     3       0
302   57    0   1       130   236    0        0      174      0      0.0      1   1     2       0

[303 rows x 14 columns]

Question 2:

Return the shape of the dataframe.

In [206]:
df.shape

(303, 14)

Question 3:

Return a Series which represent the number of people having heart disease and not having heart disease.

In [207]:
df["target"].value_counts()

1    165
0    138
Name: target, dtype: int64

Question 4:

Find the percentage of people having heart diseases. Return the percentage number with two decimal places only.

In [208]:
num_disease = len(df[df["target"]==1])
num_disease

165

In [209]:
num_nodisease = len(df[df["target"]==0])
num_nodisease

138

In [210]:
np.round(num_disease/(num_disease+num_nodisease)*100, 2)

54.46

Question 5:

Find the mean of every column for each target (having/not having heart disease) and return the result as a dataframe.

In [211]:
df.groupby("target").mean()

age       sex        cp    trestbps        chol  ...     exang   oldpeak     slope        ca      thal
target                                                         ...                                                  
0       56.601449  0.826087  0.478261  134.398551  251.086957  ...  0.550725  1.585507  1.166667  1.166667  2.543478
1       52.496970  0.563636  1.375758  129.303030  242.230303  ...  0.139394  0.583030  1.593939  0.363636  2.121212

[2 rows x 13 columns]

Question 6:

Create dummy variables for the columns 'cp', 'thal' and 'slope'. After creating the variables, concatenate those dummy variables in the original dataframe and return the modified dataframe. The shape should be (303, 25)

IMPORTANT: For grading purposes, make sure that the created dummy variables are in this order: 'cp', 'thal', 'slope'. Sample output:

In [212]:
dummy_variables = pd.get_dummies(df[["cp","thal","slope"]], columns=["cp","thal","slope"])
dummy_variables

cp_0  cp_1  cp_2  cp_3  thal_0  thal_1  thal_2  thal_3  slope_0  slope_1  slope_2
0       0     0     0     1       0       1       0       0        1        0        0
1       0     0     1     0       0       0       1       0        1        0        0
2       0     1     0     0       0       0       1       0        0        0        1
3       0     1     0     0       0       0       1       0        0        0        1
4       1     0     0     0       0       0       1       0        0        0        1
..    ...   ...   ...   ...     ...     ...     ...     ...      ...      ...      ...
298     1     0     0     0       0       0       0       1        0        1        0
299     0     0     0     1       0       0       0       1        0        1        0
300     1     0     0     0       0       0       0       1        0        1        0
301     1     0     0     0       0       0       0       1        0        1        0
302     0     1     0     0       0       0       1       0        0        1        0

[303 rows x 11 columns]

In [213]:
df = pd.concat([df, dummy_variables], axis=1)
df

age  sex  cp  trestbps  chol  fbs  restecg  ...  thal_0  thal_1  thal_2  thal_3  slope_0  slope_1  slope_2
0     63    1   3       145   233    1        0  ...       0       1       0       0        1        0        0
1     37    1   2       130   250    0        1  ...       0       0       1       0        1        0        0
2     41    0   1       130   204    0        0  ...       0       0       1       0        0        0        1
3     56    1   1       120   236    0        1  ...       0       0       1       0        0        0        1
4     57    0   0       120   354    0        1  ...       0       0       1       0        0        0        1
..   ...  ...  ..       ...   ...  ...      ...  ...     ...     ...     ...     ...      ...      ...      ...
298   57    0   0       140   241    0        1  ...       0       0       0       1        0        1        0
299   45    1   3       110   264    0        1  ...       0       0       0       1        0        1        0
300   68    1   0       144   193    1        1  ...       0       0       0       1        0        1        0
301   57    1   0       130   131    0        1  ...       0       0       0       1        0        1        0
302   57    0   1       130   236    0        0  ...       0       0       1       0        0        1        0

[303 rows x 25 columns]

Question 7:

Drop the original columns 'cp', 'thal', 'slope'. You should have 303 rows and 22 columns.

In [214]:
df.drop(columns=["cp", "thal", "slope"], axis = 1, inplace=True)

In [215]:
df

age  sex  trestbps  chol  fbs  restecg  thalach  ...  thal_0  thal_1  thal_2  thal_3  slope_0  slope_1  slope_2
0     63    1       145   233    1        0      150  ...       0       1       0       0        1        0        0
1     37    1       130   250    0        1      187  ...       0       0       1       0        1        0        0
2     41    0       130   204    0        0      172  ...       0       0       1       0        0        0        1
3     56    1       120   236    0        1      178  ...       0       0       1       0        0        0        1
4     57    0       120   354    0        1      163  ...       0       0       1       0        0        0        1
..   ...  ...       ...   ...  ...      ...      ...  ...     ...     ...     ...     ...      ...      ...      ...
298   57    0       140   241    0        1      123  ...       0       0       0       1        0        1        0
299   45    1       110   264    0        1      132  ...       0       0       0       1        0        1        0
300   68    1       144   193    1        1      141  ...       0       0       0       1        0        1        0
301   57    1       130   131    0        1      115  ...       0       0       0       1        0        1        0
302   57    0       130   236    0        0      174  ...       0       0       1       0        0        1        0

[303 rows x 22 columns]

Question 8:

Split the data into training and testing sets, with 80% training and 20% testing.

IMPORTANT: For grading purposes, please set random_state=42.

In [216]:
from sklearn.model_selection import train_test_split
df_train, df_test = train_test_split(df, test_size=0.2, random_state=42)

Question 9:

Use MinMaxScaler to scale the X_train and X_test datasets. Return X_train_scaled, X_test_scaled.

In [217]:
X_train = pd.concat([df_train.iloc[:, :10], df_train.iloc[:,11:]], axis=1).values
len(X_train)

242

In [218]:
y_train = df_train.iloc[:,10].values
len(y_train)

242

In [219]:
X_test = pd.concat([df_test.iloc[:, :10], df_test.iloc[:,11:]], axis=1).values
len(X_test)

61

In [220]:
y_test = df_test.iloc[:,10].values
len(y_test)

61

In [221]:
from sklearn.preprocessing import MinMaxScaler
sc = MinMaxScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

In [222]:
X_train

array([[0.27083333, 1.        , 0.26530612, ..., 0.        , 0.        ,
        1.        ],
       [0.60416667, 1.        , 0.57142857, ..., 0.        , 0.        ,
        1.        ],
       [0.35416667, 1.        , 0.57142857, ..., 0.        , 1.        ,
        0.        ],
       ...,
       [0.83333333, 1.        , 0.67346939, ..., 0.        , 1.        ,
        0.        ],
       [0.35416667, 1.        , 0.26530612, ..., 0.        , 0.        ,
        1.        ],
       [0.70833333, 0.        , 0.46938776, ..., 0.        , 0.        ,
        1.        ]])

In [223]:
X_test

array([[0.58333333, 1.        , 0.57142857, ..., 0.        , 1.        ,
        0.        ],
       [0.625     , 1.        , 0.7755102 , ..., 0.        , 1.        ,
        0.        ],
       [0.58333333, 1.        , 0.57142857, ..., 0.        , 0.        ,
        1.        ],
       ...,
       [0.8125    , 1.        , 0.51020408, ..., 0.        , 1.        ,
        0.        ],
       [0.64583333, 1.        , 0.52040816, ..., 0.        , 1.        ,
        0.        ],
       [0.4375    , 1.        , 0.57142857, ..., 0.        , 1.        ,
        0.        ]])

Question 10:

Build a Logistic Regression model and use that model to make predictions. Return the accuracy up to 2 decimal places (should be around 0.89) and the value of true positive.

In [224]:
from sklearn.linear_model import LogisticRegression
model_lr = LogisticRegression()
model_lr = model_lr.fit(X_train, y_train)

In [225]:
y_pred_lr = model_lr.predict(X_test)
y_pred_lr

array([0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0,
       0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0], dtype=int64)

In [226]:
y_test

array([0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0,
       0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0], dtype=int64)

In [227]:
from sklearn.metrics import accuracy_score
score_lr = accuracy_score(y_pred_lr, y_test)
print(f"The accuracy is {np.round(score_lr, 2)}")

The accuracy is 0.89


In [228]:
from sklearn.metrics import confusion_matrix
matrix_lr = confusion_matrix(y_pred_lr, y_test)
print(matrix_lr)
print(f"The true positive is {matrix_lr[0][0]}")

[[27  5]
 [ 2 27]]
The true positive is 27


Question 11:

Build a KNN with 3 neighbors model and use that model to make predictions. Return the accuracy up to 2 decimal places (should be around 0.79) and the value of true positive.

In [229]:
from sklearn.neighbors import KNeighborsClassifier
model_knn = KNeighborsClassifier(n_neighbors=3)
model_knn = model_knn.fit(X_train, y_train)

In [230]:
y_pred_knn = model_knn.predict(X_test)
y_pred_knn

C:\Users\Admin\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)


array([0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0,
       1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0,
       1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0], dtype=int64)

In [231]:
y_test

array([0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0,
       0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0], dtype=int64)

In [234]:
from sklearn.metrics import accuracy_score
score_knn = accuracy_score(y_pred_knn, y_test)
print(f"The accuracy of knn model is {np.round(score_knn,2)}")

The accuracy of knn model is 0.8


In [236]:
from sklearn.metrics import confusion_matrix
matrix_knn = confusion_matrix(y_pred_knn, y_test)
print(matrix_knn)
print(f"The true positive is {matrix_knn[0][0]}")

[[26  9]
 [ 3 23]]
The true positive is 26
